numpy/src/numpy.cpp

```diff
@@ -939,6 +1034,104 @@ using ndarray_float32 = ndarray<float32>;
 using ndarray_float = ndarray<float64>;
 using ndarray_float = ndarray<float_>;
 
+// Define template for creating n-dimensional vectors
+template <typename T, std::size_t N>
+struct nvector_impl {
+    using type = std::vector<typename nvector_impl<T, N - 1>::type>;
+};
+template <typename T>
+struct nvector_impl<T, 0> {
+    using type = T;
+};
+template <typename T, std::size_t N>
+using nvector = typename nvector_impl<T, N>::type;
+
+// Transform nvector<U, N> to nvector<T, N>
+template <typename U, typename T, std::size_t N>
+nvector<T, N> transform(const nvector<U, N>& values) {
+    nvector<T, N> result;
+    if constexpr(N != 0) {
+        for (const auto& value : values) {
+            result.push_back(transform<U, T, N - 1>(value));
+        }
+    } else {
+        result = static_cast<T>(values);
+    }
+    return result;
+}
+
+void register_array_int(py::module_& m) {
```

Comment:
Merged this for now. Will remove the single data case in a further commit, once I test it.